ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/pathbased/bonsai/BonsaiArchiveProofsWorldStateProvider.java

```diff
@@ -0,0 +1,253 @@
+    if (queryParams.shouldWorldStateUpdateHead()) {
+      return getFullWorldState(queryParams);
+    } else {
+
+      final Optional<BlockHeader> checkpointBlock =
+          getCheckpointStateStartBlock(blockchain, queryParams.getBlockHeader().getHash());
+
+      if (checkpointBlock.isPresent()) {
+        return cachedWorldStorageManager
+            .getWorldState(checkpointBlock.get().getHash())
+            .or(() -> cachedWorldStorageManager.getHeadWorldState(blockchain::getBlockHeaderSafe))
+            .flatMap(
+                worldState ->
+                    rollArchiveProofWorldStateToBlockHash(
+                        worldState, checkpointBlock.get(), queryParams.getBlockHeader().getHash()))
+            .map(MutableWorldState::freezeStorage);
+      }
+      return Optional.empty();
+    }
+  }
+
+  private Optional<MutableWorldState> rollArchiveProofWorldStateToBlockHash(
+      final PathBasedWorldState mutableState,
+      final BlockHeader checkpointBlock,
+      final Hash targetBlockHash) {
+
+    // Reset the current world state to the desired checkpoint block number by setting its metadata
+    // to the checkpoint block number & hash. We know that the archive world state is complete at
+    // every checkpoint block, so to use the state for a given checkpoint block we simply need to
+    // assert which block it represents.
+
+    // E.g. chain head = 230, target = 120, checkpoint blocks every 100 blocks. Since archive state
+    // at chainhead includes everything we need recreate state at any block we simply assert that
+    // the current world state is for block 200 (e.g. the nearest checkpoint). Rolling back will
+    // then start from 200 and replay trie logs backwards to 120. We will then PUT new trie nodes
+    // as if we were at block 120, and after those puts (during mutablestate.persist()) the
+    // WORLD_BLOCK_NUMBER_KEY, WORLD_BLOCK_HASH_KEY, and WORLD_ROOT_HASH_KEY will be set to 120.
+    ((BonsaiArchiveWorldState) mutableState).resetWorldStateToCheckpoint(checkpointBlock);
+
+    // Roll back from the checkpoint state to the target block hash
+    if (targetBlockHash.equals(mutableState.blockHash())) {
+      return Optional.of(mutableState);
+    } else {
+      try {
+        final Optional<BlockHeader> maybePersistedHeader =
+            blockchain
+                .getBlockHeader(mutableState.blockHash())
+                .or(() -> blockchain.getBlockHeaderSafe(mutableState.blockHash()))
+                .map(BlockHeader.class::cast);
+
+        final List<TrieLog> rollBacks = new ArrayList<>();
+        if (maybePersistedHeader.isEmpty()) {
+          trieLogManager.getTrieLogLayer(mutableState.blockHash()).ifPresent(rollBacks::add);
+        } else {
+          BlockHeader targetHeader =
+              blockchain
+                  .getBlockHeader(targetBlockHash)
+                  .or(() -> blockchain.getBlockHeaderSafe(targetBlockHash))
+                  .get();
```

Comment:
is this safe? I don't think we can guarantee we will get the block header even with `getBlockHeaderSafe`